File: android/src/main/java/com/urbanairship/reactnative/ReactMessageView.java
/* Copyright Airship and Contributors */

package com.urbanairship.reactnative;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.urbanairship.Cancelable;
import com.urbanairship.UAirship;
import com.urbanairship.richpush.RichPushInbox;
import com.urbanairship.richpush.RichPushMessage;
import com.urbanairship.widget.UAWebView;
import com.urbanairship.widget.UAWebViewClient;

public class ReactMessageView extends FrameLayout implements LifecycleEventListener {

    static final String EVENT_LOAD_STARTED = "onLoadStarted";
    static final String EVENT_LOAD_FINISHED = "onLoadFinished";
    static final String EVENT_LOAD_ERROR = "onLoadError";
    static final String EVENT_CLOSE = "onClose";

    private static final String MESSAGE_ID_KEY = "messageId";
    private static final String RETRYABLE_KEY = "retryable";
    private static final String ERROR_KEY = "error";

    private static final String ERROR_MESSAGE_NOT_AVAILABLE = "MESSAGE_NOT_AVAILABLE";
    private static final String ERROR_FAILED_TO_FETCH_MESSAGE = "FAILED_TO_FETCH_MESSAGE";
    private static final String ERROR_MESSAGE_LOAD_FAILED = "MESSAGE_LOAD_FAILED";

    private RichPushMessage message;
    private Cancelable fetchMessageRequest;
    private UAWebView webView;

    private WebViewClient webViewClient = new UAWebViewClient() {
        private Integer error = null;

        @Override
        public void onPageFinished(WebView view, String url) {
            super.onPageFinished(view, url);

            if (message == null) {
                return;
            }

            if (error != null) {
                notifyLoadError(message.getMessageId(), ERROR_MESSAGE_LOAD_FAILED, false);
                return;
            }

            message.markRead();
            notifyLoadFinished(message.getMessageId());
        }

        @Override
        public void onReceivedError(WebView view, int errorCode, String description, @Nullable String failingUrl) {
            super.onReceivedError(view, errorCode, description, failingUrl);

            if (message != null && failingUrl != null && failingUrl.equals(message.getMessageBodyUrl())) {
                error = errorCode;
            }
        }

        @Override
        public void onClose(@NonNull WebView webView) {
            if (message != null) {
                notifyClose(message.getMessageId());
            }
        }
    };

    public ReactMessageView(@NonNull Context context) {
        super(context);
    }

    public void loadMessage(final String messageId) {
        if (webView == null) {
            webView = new UAWebView(getContext());
            webView.setWebViewClient(webViewClient);
            addView(webView);
        }

        if (fetchMessageRequest != null) {
            fetchMessageRequest.cancel();
        }
        this.message = null;
        startLoading(messageId);
    }

    void startLoading(final String messageId) {
        notifyLoadStarted(messageId);

        this.message = UAirship.shared().getInbox().getMessage(messageId);
        if (this.message == null) {
            fetchMessageRequest = UAirship.shared().getInbox().fetchMessages(new RichPushInbox.FetchMessagesCallback() {
                @Override
                public void onFinished(boolean success) {
                    RichPushMessage message = UAirship.shared().getInbox().getMessage(messageId);
                    if (!success) {
                        notifyLoadError(messageId, ERROR_FAILED_TO_FETCH_MESSAGE, true);
                        return;
                    } else if (message == null || message.isExpired()) {
                        notifyLoadError(messageId, ERROR_MESSAGE_NOT_AVAILABLE, false);
                        return;
                    }
                    webView.loadRichPushMessage(message);
                }
            });
        } else {
            if (this.message.isExpired()) {
                notifyLoadError(messageId, ERROR_MESSAGE_NOT_AVAILABLE, false);
                return;
            }
            webView.loadRichPushMessage(this.message);
        }
    }

    private void notifyLoadError(String messageId, String error, boolean retryable) {
        WritableMap event = Arguments.createMap();
        event.putString(MESSAGE_ID_KEY, messageId);
        event.putBoolean(RETRYABLE_KEY, retryable);
        event.putString(ERROR_KEY, error);
        notify(EVENT_LOAD_ERROR, event);
    }

    private void notifyLoadFinished(String messageId) {
        WritableMap event = Arguments.createMap();
        event.putString(MESSAGE_ID_KEY, messageId);
        notify(EVENT_LOAD_FINISHED, event);
    }

    private void notifyLoadStarted(String messageId) {
        WritableMap event = Arguments.createMap();
        event.putString(MESSAGE_ID_KEY, messageId);
        notify(EVENT_LOAD_STARTED, event);
    }

    private void notifyClose(String messageId) {
        WritableMap event = Arguments.createMap();
        event.putString(MESSAGE_ID_KEY, messageId);
        notify(EVENT_CLOSE, event);
    }

    private void notify(String eventName, WritableMap event) {
        ReactContext reactContext = (ReactContext) getContext();
        reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                getId(),
                eventName,
                event);
    }

    @Override
    public void onHostResume() {
        if (webView != null) {
            webView.onResume();
        }
    }

    @Override
    public void onHostPause() {
        if (webView != null) {
            webView.onPause();
        }
    }

    @Override
    public void onHostDestroy() {
        cleanup();
    }

    public void cleanup() {
        if (webView != null) {
            webView.setWebViewClient(null);
            webView.destroy();
            webView = null;
        }
    }
}

File: android/src/main/java/com/urbanairship/reactnative/ReactMessageView.kt
/* Copyright Airship and Contributors */

package com.urbanairship.reactnative

import android.content.Context
import android.view.ContextThemeWrapper
import android.widget.FrameLayout
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.urbanairship.messagecenter.Message
import com.urbanairship.messagecenter.ui.view.MessageView
import com.urbanairship.messagecenter.ui.view.MessageViewState
import com.urbanairship.messagecenter.ui.view.MessageViewModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus

class ReactMessageView(context: Context) : FrameLayout(context), LifecycleEventListener {

    private val viewModel = MessageViewModel()
    private val messageView: MessageView = MessageView(
        ContextThemeWrapper(context, R.style.RNAirshipMessageViewTheme)
    )
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main.immediate) + SupervisorJob()
    private var currentMessageId: String? = null

    // React Native intercepts requestLayout and doesn't always propagate it to children.
    override fun requestLayout() {
        super.requestLayout()
        post {
            measure(
                MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
            )
            layout(left, top, right, bottom)
        }
    }

    init {
        addView(messageView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))

        messageView.listener = object : MessageView.Listener {
            override fun onMessageLoaded(message: Message) {
                viewModel.markMessagesRead(message)
                notifyLoadFinished(message.id)
            }

            override fun onMessageLoadError(error: MessageViewState.Error.Type) {
                val messageId = currentMessageId ?: return
                // If the ViewModel is in an error state, the error originated from fetching.
                // Otherwise it came from the WebView itself.
                val (errorStr, retryable) = if (viewModel.states.value is MessageViewState.Error) {
                    when (error) {
                        MessageViewState.Error.Type.LOAD_FAILED -> ERROR_FAILED_TO_FETCH_MESSAGE to true
                        MessageViewState.Error.Type.UNAVAILABLE -> ERROR_MESSAGE_NOT_AVAILABLE to false
                    }
                } else {
                    ERROR_MESSAGE_LOAD_FAILED to false
                }
                notifyLoadError(messageId, errorStr, retryable)
            }

            override fun onRetryClicked() {
                currentMessageId?.let { viewModel.loadMessage(it) }
            }

            override fun onCloseMessage() {
                currentMessageId?.let { notifyClose(it) }
            }
        }

        scope.launch {
            viewModel.states.collect { state ->
                messageView.render(state)
                if (state is MessageViewState.Loading) {
                    currentMessageId?.let { notifyLoadStarted(it) }
                }
            }
        }
    }

    fun loadMessage(messageId: String) {
        currentMessageId = messageId
        viewModel.loadMessage(messageId)
    }

    private fun notifyLoadError(messageId: String, error: String, retryable: Boolean) {
        val event = Arguments.createMap()
        event.putString(MESSAGE_ID_KEY, messageId)
        event.putBoolean(RETRYABLE_KEY, retryable)
        event.putString(ERROR_KEY, error)
        notify(EVENT_LOAD_ERROR, event)
    }

    private fun notifyLoadFinished(messageId: String) {
        val event = Arguments.createMap()
        event.putString(MESSAGE_ID_KEY, messageId)
        notify(EVENT_LOAD_FINISHED, event)
    }

    private fun notifyLoadStarted(messageId: String) {
        val event = Arguments.createMap()
        event.putString(MESSAGE_ID_KEY, messageId)
        notify(EVENT_LOAD_STARTED, event)
    }

    private fun notifyClose(messageId: String) {
        val event = Arguments.createMap()
        event.putString(MESSAGE_ID_KEY, messageId)
        notify(EVENT_CLOSE, event)
    }

    private fun notify(eventName: String, event: WritableMap) {
        val reactContext = context as ReactContext
        reactContext.airshipDispatchEvent(id, eventName, event)
    }

    override fun onHostResume() {
        messageView.resumeWebView()
    }

    override fun onHostPause() {
        messageView.pauseWebView()
    }

    override fun onHostDestroy() {
        cleanup()
    }

    fun cleanup() {
        scope.cancel()
        viewModel.clearMessage()
    }

    companion object {
        const val EVENT_LOAD_STARTED = "topAirshipMessageViewLoadStarted"
        const val EVENT_LOAD_FINISHED = "topAirshipMessageViewLoadFinished"
        const val EVENT_LOAD_ERROR = "topAirshipMessageViewLoadError"
        const val EVENT_CLOSE = "topAirshipMessageViewClose"

        const val EVENT_LOAD_STARTED_HANDLER_NAME = "onLoadStarted"
        const val EVENT_LOAD_FINISHED_HANDLER_NAME = "onLoadFinished"
        const val EVENT_LOAD_ERROR_HANDLER_NAME = "onLoadError"
        const val EVENT_CLOSE_HANDLER_NAME = "onClose"

        private const val MESSAGE_ID_KEY = "messageId"
        private const val RETRYABLE_KEY = "retryable"
        private const val ERROR_KEY = "error"

        private const val ERROR_MESSAGE_NOT_AVAILABLE = "MESSAGE_NOT_AVAILABLE"
        private const val ERROR_FAILED_TO_FETCH_MESSAGE = "FAILED_TO_FETCH_MESSAGE"
        private const val ERROR_MESSAGE_LOAD_FAILED = "MESSAGE_LOAD_FAILED"
    }
}
